remotes/docker/authorizer.go

```diff
@@ -109,15 +121,29 @@ func (a *dockerAuthorizer) Authorize(ctx context.Context, req *http.Request) err
 		return nil
 	}
 
-	auth, err := ah.authorize(ctx)
+	auth, refreshToken, err := ah.authorize(ctx)
 	if err != nil {
 		return err
 	}
 
 	req.Header.Set("Authorization", auth)
+	if refreshToken != "" {
+		a.mu.Lock()
+		a.fetchedRefreshToken = refreshToken
+		a.mu.Unlock()
+	}
 	return nil
 }
 
+// RefreshToken returns a refresh token.
+// Needs to be called after Authorize().
+func (a *dockerAuthorizer) RefreshToken() string {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+
+	return a.fetchedRefreshToken
```

Comment:
`dockerAuthorizer` seems to designed to handle multiple hosts so `fetchedRefreshToken` should be reset everytime a method of `dockerAuthorizer` is called (`Authorize()`, `AddResponses()`) to avoid returning the previous result.

Reply:
fixed to use a callback function

```go
// OnFetchRefreshToken is called on fetching request token.
type OnFetchRefreshToken func(ctx context.Context, refreshToken string, req *http.Request)

// WithFetchRefreshToken enables fetching "refresh token" (aka "identity token", "offline token").
func WithFetchRefreshToken(f OnFetchRefreshToken) AuthorizerOpt {
       return func(opt *authorizerConfig) {
               opt.onFetchRefreshToken = f
       }
}
```

Reply:
Will the callback function be called concurrently by multiple goroutines? If true, may need document that should handle race condition in the callback function.